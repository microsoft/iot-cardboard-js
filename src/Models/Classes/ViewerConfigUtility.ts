import {
    IScenesConfig,
    IBehavior,
    IScene,
    DatasourceType,
    ITwinToObjectMapping
} from './3DVConfig';

/** Static utilty methods for operations on the ViewerConfiguration file. */
abstract class ViewerConfigUtility {
    /** Add new scene to config file */
    static addScene(config: IScenesConfig, scene: IScene): IScenesConfig {
        const updatedConfig = { ...config };
        updatedConfig.viewerConfiguration.scenes.push(scene);
        return updatedConfig;
    }

    /** Update scene with matching ID */
    static editScene(
        config: IScenesConfig,
        sceneId: string,
        scene: IScene
    ): IScenesConfig {
        const sceneIndex: number = config.viewerConfiguration.scenes.findIndex(
            (s) => s.id === sceneId
        );
        const updatedConfig = { ...config };
        updatedConfig.viewerConfiguration.scenes[sceneIndex] = scene;

        return updatedConfig;
    }

    /** Delete scene with matching ID */
    static deleteScene(config: IScenesConfig, sceneId: string): IScenesConfig {
        const sceneIndex: number = config.viewerConfiguration.scenes.findIndex(
            (s) => s.id === sceneId
        );
        const updatedConfig = { ...config };
        updatedConfig.viewerConfiguration.scenes.splice(sceneIndex, 1);
        return updatedConfig;
    }

    /** Add behavior to target scene */
    static addBehavior(
        config: IScenesConfig,
        sceneId: string,
        behavior: IBehavior
    ): IScenesConfig {
        const updatedConfig = { ...config };
        updatedConfig.viewerConfiguration.behaviors.push(behavior);
        updatedConfig.viewerConfiguration.scenes
            .find((scene) => scene.id === sceneId)
            ?.behaviors?.push(behavior.id);
        return updatedConfig;
    }

    /** Update behavior with matching ID.
     * Note, original behavior ID is used in case the behavior ID has been
     * changed with the update.*/
    static editBehavior(
        config: IScenesConfig,
        behavior: IBehavior,
        originalBehaviorId: string
    ): IScenesConfig {
        const updatedConfig = { ...config };

        // Update modified behavior
        const behaviorIdx = updatedConfig.viewerConfiguration.behaviors.findIndex(
            (b) => b.id === originalBehaviorId
        );
        updatedConfig.viewerConfiguration.behaviors[behaviorIdx] = behavior;

        // If behavior ID changed, update matching scene behavior Ids with updated Id
        if (behavior.id !== originalBehaviorId) {
            updatedConfig.viewerConfiguration.scenes.forEach((scene) => {
                const behaviorIdIdxInSceneBehaviors = scene?.behaviors?.indexOf(
                    originalBehaviorId
                );
                if (behaviorIdIdxInSceneBehaviors !== -1) {
                    scene.behaviors[behaviorIdIdxInSceneBehaviors] =
                        behavior.id;
                }
            });
        }

        return updatedConfig;
    }

    /** Adds existing behavior to the target scene */
    static addBehaviorToScene(
        config: IScenesConfig,
        sceneId: string,
        behavior: IBehavior
    ): IScenesConfig {
        const updatedConfig = { ...config };
        updatedConfig.viewerConfiguration.scenes
            .find((scene) => scene.id === sceneId)
            ?.behaviors?.push(behavior.id);
        return updatedConfig;
    }

    /** Delete behavior.
     * Options for deletion from current scene, or ALL scenes.
     */
    static deleteBehavior(
        config: IScenesConfig,
        sceneId: string,
        behaviorId: string,
        removeFromAllScenes?: boolean
    ): IScenesConfig {
        const updatedConfig = { ...config };

        // Remove behavior from active scene
        const activeScene = updatedConfig.viewerConfiguration.scenes.find(
            (scene) => scene.id === sceneId
        );

        const matchingBehaviorIdxInActiveScene = activeScene.behaviors.indexOf(
            behaviorId
        );

        if (matchingBehaviorIdxInActiveScene !== -1) {
            activeScene.behaviors.splice(matchingBehaviorIdxInActiveScene, 1);
        }

        // Clean up behavior datasources when removing behavior from scene reference only
        const elementIdsInActiveScene = ViewerConfigUtility.getDictionaryOfElementsIdsInScene(
            config,
            sceneId
        );

        const behavior = config.viewerConfiguration.behaviors.find(
            (b) => b.id === behaviorId
        );

        if (behavior) {
            const twinToObjectMapping = behavior.datasources.find(
                (ds) => ds.type === DatasourceType.TwinToObjectMapping
            );
            if (twinToObjectMapping) {
                twinToObjectMapping.mappingIDs = twinToObjectMapping.mappingIDs.filter(
                    (id) => !(id in elementIdsInActiveScene)
                );
            }
        }

        if (removeFromAllScenes) {
            // Splice behavior out of behavior list
            const behaviorIdx = updatedConfig.viewerConfiguration.behaviors.findIndex(
                (b) => b.id === behaviorId
            );

            if (behaviorIdx !== -1) {
                updatedConfig.viewerConfiguration.behaviors.splice(
                    behaviorIdx,
                    1
                );
            }

            // If matching behavior Id found in ANY scene, splice out scene's behavior Id array
            updatedConfig.viewerConfiguration.scenes.forEach((scene) => {
                const matchingBehaviorIdIdx = scene.behaviors.indexOf(
                    behaviorId
                );
                if (matchingBehaviorIdIdx !== -1) {
                    scene.behaviors.splice(matchingBehaviorIdIdx, 1);
                }
            });
        }
        return updatedConfig;
    }

    static getBehaviorElementIds(behavior: IBehavior) {
        return (
            behavior.datasources.find(
                (ds) => ds.type === DatasourceType.TwinToObjectMapping
            )?.mappingIDs || []
        );
    }

    /** Returns information about the number of element and scene references on a behavior  */
    static getBehaviorMetaData(
        config: IScenesConfig,
        sceneId: string,
        behavior: IBehavior
    ): { numElementsInActiveScene: number; numSceneRefs: number } {
        const dictionaryOfElementsIdsInScene = ViewerConfigUtility.getDictionaryOfElementsIdsInScene(
            config,
            sceneId
        );

        let numElementsInActiveScene = 0;

        const behaviorElementIds = ViewerConfigUtility.getBehaviorElementIds(
            behavior
        );
        behaviorElementIds.forEach((elementId) => {
            if (elementId in dictionaryOfElementsIdsInScene) {
                numElementsInActiveScene++;
            }
        });

        let numSceneRefs = 0;

        config.viewerConfiguration.scenes.forEach((s) => {
            if (s.behaviors.includes(behavior.id)) {
                numSceneRefs++;
            }
        });

        return {
            numElementsInActiveScene,
            numSceneRefs
        };
    }

    /** Returns information about the number of behaviors and meshes on a element  */
    static getElementMetaData(
        element: ITwinToObjectMapping,
        config: IScenesConfig
    ): { numBehaviors: number; numMeshes: number } {
        const numMeshes = element.meshIDs?.length;
        const numBehaviors = this.getBehaviorsOnElement(
            element,
            config?.viewerConfiguration?.behaviors
        )?.length;
        return {
            numBehaviors,
            numMeshes
        };
    }

    static getDictionaryOfElementsIdsInScene(
        config: IScenesConfig,
        sceneId: string
    ): Record<string, any> {
        // Build up dictionary of all active element IDs on current scene
        const scene = config.viewerConfiguration.scenes?.find(
            (s) => s.id === sceneId
        );
        const elementIdMap = {};

        scene?.twinToObjectMappings?.forEach((ttom) => {
            if (!(ttom.id in elementIdMap)) elementIdMap[ttom.id] = ttom;
        });

        return elementIdMap;
    }

    static getBehaviorsSegmentedByPresenceInScene(
        config: IScenesConfig,
        sceneId: string,
        behaviors: Array<IBehavior>
    ): [
        behaviorsInScene: Array<IBehavior>,
        behaviorsNotInScene: Array<IBehavior>
    ] {
        const behaviorsInScene = [];
        const behaviorsNotInScene = [];

        const scene = config.viewerConfiguration.scenes.find(
            (s) => s.id === sceneId
        );

        const behaviorIdsInActiveScene = scene?.behaviors;

        behaviors.forEach((behavior) => {
            if (
                behaviorIdsInActiveScene &&
                behaviorIdsInActiveScene.includes(behavior.id)
            ) {
                behaviorsInScene.push(behavior);
            } else {
                behaviorsNotInScene.push(behavior);
            }
        });

        // Sanity check to ensure all behaviors have been segmented
        if (
            behaviorsInScene.length + behaviorsNotInScene.length ===
            behaviors.length
        ) {
            return [behaviorsInScene, behaviorsNotInScene];
        } else {
            return [[], behaviors];
        }
    }

    static getBehaviorsOnElement(
        element: ITwinToObjectMapping,
        behaviors: Array<IBehavior>
    ) {
        return (
            behaviors?.filter((behavior) =>
                behavior.datasources?.[0]?.mappingIDs?.find(
                    (id) => id === element?.id
                )
            ) || []
        );
    }

    static getMappingIdsForBehavior(behavior: IBehavior) {
        const mappingIds: string[] = [];
        // cycle through the datasources of behavior
        for (const dataSource of behavior.datasources) {
            // if its a TwinToObjectMappingDatasource get the mapping id
            if (dataSource.type === DatasourceType.TwinToObjectMapping) {
                dataSource.mappingIDs.forEach((mappingId) => {
                    mappingIds.push(mappingId);
                });
            }
        }

        return mappingIds;
    }
}

export default ViewerConfigUtility;
