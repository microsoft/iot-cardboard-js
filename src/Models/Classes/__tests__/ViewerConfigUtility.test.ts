import { cleanup } from '@testing-library/react-hooks';
import {
    DTwin,
    DEFAULT_REFRESH_RATE_IN_MILLISECONDS,
    PRIMARY_TWIN_NAME
} from '../../Constants';
import {
    I3DScenesConfig,
    IPollingConfiguration,
    IScene
} from '../../Types/Generated/3DScenesConfiguration-v1.0.0';
import ViewerConfigUtility from '../ViewerConfigUtility';
import {
    ALIASED_TWIN_ID_1,
    ALIASED_TWIN_ID_2,
    CUSTOM_ELEMENT,
    getTwinWithInFlowOutFlow,
    getTwinWithInFlowPercentFull,
    getTwinWithTemperature,
    MOCK_BEHAVIOR_1,
    MOCK_BEHAVIOR_2,
    MOCK_BEHAVIOR_3,
    MOCK_ELEMENT_1,
    MOCK_ELEMENT_2,
    MOCK_ELEMENT_3_WITH_ALIASES
} from './ViewerConfigUtilityTestHelpers';

afterEach(cleanup);

describe('ViewerConfigUtility', () => {
    describe('Scene Visuals', () => {
        const MOCK_SCENE_ID = 'SceneId1';
        /**
         * Has 3 behaviors, only 2 are used in the scene (1,2)
         * Has 3 elements, 2 are real and one is custom so we should ignore it
         * @returns mock data for the default scene
         */
        const getMockConfig = (): I3DScenesConfig => {
            return {
                $schema: {} as any,
                configuration: {
                    behaviors: [
                        MOCK_BEHAVIOR_1,
                        MOCK_BEHAVIOR_2,
                        MOCK_BEHAVIOR_3
                    ],
                    layers: [],
                    scenes: [
                        {
                            id: MOCK_SCENE_ID,
                            displayName: 'Mock display name',
                            behaviorIDs: [
                                MOCK_BEHAVIOR_1.id,
                                MOCK_BEHAVIOR_2.id
                            ],
                            elements: [
                                MOCK_ELEMENT_1,
                                MOCK_ELEMENT_3_WITH_ALIASES,
                                CUSTOM_ELEMENT
                            ],
                            assets: []
                        }
                    ]
                }
            };
        };
        /** create twin data for 3 elements, the 3rd has aliases which get data for the same properties as element 1 */
        const getMockTwinData = (): Map<string, DTwin> => {
            const data = new Map<string, DTwin>();
            data.set(
                MOCK_ELEMENT_1.primaryTwinID,
                getTwinWithInFlowOutFlow(MOCK_ELEMENT_1.primaryTwinID)
            );
            data.set(
                MOCK_ELEMENT_2.primaryTwinID,
                getTwinWithInFlowPercentFull(MOCK_ELEMENT_2.primaryTwinID)
            );
            data.set(
                MOCK_ELEMENT_3_WITH_ALIASES.primaryTwinID,
                getTwinWithTemperature(
                    MOCK_ELEMENT_3_WITH_ALIASES.primaryTwinID
                )
            );
            data.set(
                ALIASED_TWIN_ID_1,
                getTwinWithInFlowOutFlow(ALIASED_TWIN_ID_1)
            );
            data.set(
                ALIASED_TWIN_ID_2,
                getTwinWithInFlowOutFlow(ALIASED_TWIN_ID_1)
            );

            return data;
        };

        test('returns [] when null config', () => {
            // ARRANGE
            const config = undefined;
            const sceneId = MOCK_SCENE_ID;
            const twinData = getMockTwinData();

            // ACT
            const result = ViewerConfigUtility.getSceneVisualsInScene(
                config,
                sceneId,
                twinData
            );

            // ASSERT
            expect(result).toEqual([]);
        });
        test('returns [] when empty sceneId', () => {
            // ARRANGE
            const config = getMockConfig();
            const sceneId = undefined;
            const twinData = getMockTwinData();

            // ACT
            const result = ViewerConfigUtility.getSceneVisualsInScene(
                config,
                sceneId,
                twinData
            );

            // ASSERT
            expect(result).toEqual([]);
        });
        test('returns [] when scene not found', () => {
            // ARRANGE
            const config = getMockConfig();
            const sceneId = 'Some scene id that does not exist';
            const twinData = getMockTwinData();

            // ACT
            const result = ViewerConfigUtility.getSceneVisualsInScene(
                config,
                sceneId,
                twinData
            );

            // ASSERT
            expect(result).toEqual([]);
        });
        test('returns visuals when valid inputs', () => {
            // ARRANGE
            const config = getMockConfig();
            const sceneId = MOCK_SCENE_ID;
            const twinData = getMockTwinData();

            // ACT
            const result = ViewerConfigUtility.getSceneVisualsInScene(
                config,
                sceneId,
                twinData
            );

            // ASSERT
            expect(result).not.toBeFalsy();
            expect(result.length).toEqual(2); // cause 2 real elements, skips the custom
            expect(result).toMatchSnapshot(); // validate the full output

            // validate Visual 1
            const visual1 = result[0];
            expect(visual1.behaviors.length).toEqual(2); // first element should have 2 behaviors
            expect(
                visual1.behaviors.find((x) => x.id === MOCK_BEHAVIOR_1.id)
            ).toBeTruthy();
            expect(
                visual1.behaviors.find((x) => x.id === MOCK_BEHAVIOR_2.id)
            ).toBeTruthy();
            expect(visual1.coloredMeshItems).toBeUndefined(); // not populated by the helper
            expect(visual1.twins).not.toBeUndefined();
            expect(Object.keys(visual1.twins).length).toEqual(1);
            expect(visual1.twins[PRIMARY_TWIN_NAME]).not.toBeUndefined();

            // validate Visual 2
            const visual2 = result[1];
            expect(visual2.behaviors.length).toEqual(1); // second element has 2 other behaviors
            expect(
                visual2.behaviors.find((x) => x.id === MOCK_BEHAVIOR_2.id)
            ).toBeTruthy();
            expect(visual2.coloredMeshItems).toBeUndefined(); // not populated by the helper
            expect(visual2.twins).not.toBeUndefined();
            expect(Object.keys(visual2.twins).length).toEqual(3);
            expect(visual2.twins[PRIMARY_TWIN_NAME]).not.toBeUndefined();
            expect(visual2.twins['TestAlias1']).not.toBeUndefined(); // should resolve the aliases
            expect(visual2.twins['TestAlias2']).not.toBeUndefined(); // should resolve the aliases
        });
        test('returns visuals even when no elements are part of behaviors (ie. no twin data)', () => {
            // ARRANGE
            const config = getMockConfig();
            const sceneId = MOCK_SCENE_ID;
            const twinData = new Map<string, DTwin>();

            // remove all elements from behaviors just to simulate the real thing (shouldn't actually matter)
            const behaviorIds = config.configuration.scenes.find(
                (x) => x.id === sceneId
            ).behaviorIDs;
            config.configuration.behaviors
                .filter((x) => behaviorIds.includes(x.id))
                .forEach((x) => (x.datasources = []));

            // ACT
            const result = ViewerConfigUtility.getSceneVisualsInScene(
                config,
                sceneId,
                twinData
            );

            // ASSERT
            expect(result).not.toBeFalsy();
            expect(result.length).toEqual(2); // cause 2 real elements, skips the custom
            expect(result).toMatchSnapshot(); // validate the full output
        });
    });

    describe('Polling Configuration', () => {
        const TEST_SCENE_ID = 'mock_scene_id';
        const GET_TEST_SCENE_CONFIG = (
            pollingConfiguration: IPollingConfiguration
        ): I3DScenesConfig => ({
            $schema: {} as any,
            configuration: {
                scenes: [
                    {
                        id: TEST_SCENE_ID,
                        pollingConfiguration: pollingConfiguration,
                        ...({} as IScene)
                    } as IScene
                ],
                behaviors: [],
                layers: []
            }
        });

        describe('getPollingConfig', () => {
            test('returns default config when input Config is null', () => {
                // ARRANGE
                // ACT
                const result = ViewerConfigUtility.getPollingConfig(
                    null,
                    TEST_SCENE_ID
                );

                // ASSERT
                expect(result).toBeDefined();
                expect(result.minimumPollingFrequency).toEqual(
                    DEFAULT_REFRESH_RATE_IN_MILLISECONDS
                );
            });
            test('returns default config when input SceneId is null', () => {
                // ARRANGE
                // ACT
                const result = ViewerConfigUtility.getPollingConfig(
                    GET_TEST_SCENE_CONFIG({ minimumPollingFrequency: 2000 }),
                    null
                );

                // ASSERT
                expect(result).toBeDefined();
                expect(result.minimumPollingFrequency).toEqual(
                    DEFAULT_REFRESH_RATE_IN_MILLISECONDS
                );
            });
            test('returns default config when input SceneId not found in config', () => {
                // ARRANGE
                // ACT
                const result = ViewerConfigUtility.getPollingConfig(
                    GET_TEST_SCENE_CONFIG({ minimumPollingFrequency: 2000 }),
                    'non-existent-scene-id'
                );

                // ASSERT
                expect(result).toBeDefined();
                expect(result.minimumPollingFrequency).toEqual(
                    DEFAULT_REFRESH_RATE_IN_MILLISECONDS
                );
            });
            test('returns the polling configuration in config scene is found', () => {
                // ARRANGE
                // ACT
                const result = ViewerConfigUtility.getPollingConfig(
                    GET_TEST_SCENE_CONFIG({
                        minimumPollingFrequency: 1000
                    }),
                    TEST_SCENE_ID
                );

                // ASSERT
                expect(result).toBeDefined();
                expect(result.minimumPollingFrequency).toEqual(1000);
            });
        });

        describe('setPollingRate', () => {
            test('returns false when input Config is null', () => {
                // ARRANGE
                const newPollingRate = 1000;

                // ACT
                const result = ViewerConfigUtility.setPollingRate(
                    null,
                    TEST_SCENE_ID,
                    newPollingRate
                );

                // ASSERT
                expect(result).toBeFalsy();
            });
            test('returns false when input SceneId is null', () => {
                // ARRANGE
                const newPollingRate = 1000;
                const config = GET_TEST_SCENE_CONFIG({
                    minimumPollingFrequency: 2000
                });

                // ACT
                const result = ViewerConfigUtility.setPollingRate(
                    config,
                    null,
                    newPollingRate
                );

                // ASSERT
                expect(result).toBeFalsy();
            });
            test('returns false when Scene does not exist is null', () => {
                // ARRANGE
                const newPollingRate = 1000;
                const config = GET_TEST_SCENE_CONFIG({
                    minimumPollingFrequency: 2000
                });

                // ACT
                const result = ViewerConfigUtility.setPollingRate(
                    config,
                    'non-existent-scene-id',
                    newPollingRate
                );

                // ASSERT
                expect(result).toBeFalsy();
            });
            test('returns true & sets frequency to provided value if not set', () => {
                // ARRANGE
                const newPollingRate = 1000;
                const config = GET_TEST_SCENE_CONFIG({
                    minimumPollingFrequency: 5000
                });

                // ACT
                const result = ViewerConfigUtility.setPollingRate(
                    config,
                    TEST_SCENE_ID,
                    newPollingRate
                );

                // ASSERT
                expect(result).toBeTruthy();
                const newConfig = ViewerConfigUtility.getPollingConfig(
                    config,
                    TEST_SCENE_ID
                );
                expect(newConfig.minimumPollingFrequency).toEqual(
                    newPollingRate
                );
            });
        });
    });
});
