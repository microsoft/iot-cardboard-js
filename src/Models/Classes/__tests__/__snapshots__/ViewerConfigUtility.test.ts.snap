// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewerConfigUtility Scene Visuals returns visuals even when no elements are part of behaviors (ie. no twin data) 1`] = `
Array [
  SceneVisual {
    "behaviors": Array [],
    "coloredMeshItems": undefined,
    "element": Object {
      "displayName": "MockElement1",
      "id": "MockElementId1",
      "objectIDs": Array [
        "elementObject1",
        "elementObject2",
      ],
      "primaryTwinID": "mockElementTwinId1",
      "type": "TwinToObjectMapping",
    },
    "twins": Object {
      "PrimaryTwin": undefined,
    },
  },
  SceneVisual {
    "behaviors": Array [],
    "coloredMeshItems": undefined,
    "element": Object {
      "displayName": "MockElement3",
      "id": "MockElementId3",
      "objectIDs": Array [
        "elementObject2",
        "elementObject3",
      ],
      "primaryTwinID": "mockElementTwinId3",
      "twinAliases": Object {
        "TestAlias1": "aliasedTwinId1",
        "TestAlias2": "aliasedTwinId2",
      },
      "type": "TwinToObjectMapping",
    },
    "twins": Object {
      "PrimaryTwin": undefined,
      "TestAlias1": undefined,
      "TestAlias2": undefined,
    },
  },
]
`;

exports[`ViewerConfigUtility Scene Visuals returns visuals when valid inputs 1`] = `
Array [
  SceneVisual {
    "behaviors": Array [
      Object {
        "datasources": Array [
          Object {
            "elementIDs": Array [
              "MockElementId1",
            ],
            "type": "ElementTwinToObjectMappingDataSource",
          },
        ],
        "displayName": "Mock behavior 1",
        "id": "mockBehavior1",
        "visuals": Array [
          Object {
            "objectIDs": Object {
              "expression": "N/A",
            },
            "title": "Pop over visual 1",
            "type": "Popover",
            "widgets": Array [
              Object {
                "id": "widgetId1",
                "type": "Gauge",
                "valueExpression": "PrimaryTwin.testProperty",
                "widgetConfiguration": Object {
                  "label": "My widget 1",
                  "valueRanges": Array [
                    Object {
                      "id": "valuerange1",
                      "values": Array [
                        0,
                        100,
                      ],
                      "visual": Object {
                        "color": "blue",
                      },
                    },
                  ],
                },
              },
            ],
          },
          Object {
            "expressionType": "CategoricalValues",
            "objectIDs": Object {
              "expression": "N/A",
            },
            "type": "ExpressionRangeVisual",
            "valueExpression": "PrimaryTwin.Temperature",
            "valueRanges": Array [
              Object {
                "id": "valuerange1",
                "values": Array [
                  true,
                ],
                "visual": Object {
                  "color": "#d6940f",
                  "iconName": "Frigid",
                  "labelExpression": "Temp too hot",
                },
              },
            ],
          },
        ],
      },
      Object {
        "datasources": Array [
          Object {
            "elementIDs": Array [
              "MockElementId1",
              "MockElementId2",
              "MockElementId3",
            ],
            "type": "ElementTwinToObjectMappingDataSource",
          },
        ],
        "displayName": "Mock behavior 2",
        "id": "mockBehavior2",
        "visuals": Array [
          Object {
            "expressionType": "CategoricalValues",
            "objectIDs": Object {
              "expression": "N/A",
            },
            "type": "ExpressionRangeVisual",
            "valueExpression": "PrimaryTwin.Temperature",
            "valueRanges": Array [
              Object {
                "id": "valuerange1",
                "values": Array [
                  true,
                ],
                "visual": Object {
                  "color": "#d6940f",
                  "iconName": "Frigid",
                  "labelExpression": "Temp too hot",
                },
              },
            ],
          },
          Object {
            "expressionType": "NumericRange",
            "objectIDs": Object {
              "expression": "N/A",
            },
            "type": "ExpressionRangeVisual",
            "valueExpression": "PrimaryTwin.Temperature",
            "valueRanges": Array [
              Object {
                "id": "valuerange1",
                "values": Array [
                  0,
                  100,
                ],
                "visual": Object {
                  "color": "blue",
                },
              },
            ],
          },
        ],
      },
    ],
    "coloredMeshItems": undefined,
    "element": Object {
      "displayName": "MockElement1",
      "id": "MockElementId1",
      "objectIDs": Array [
        "elementObject1",
        "elementObject2",
      ],
      "primaryTwinID": "mockElementTwinId1",
      "type": "TwinToObjectMapping",
    },
    "twins": Object {
      "PrimaryTwin": Object {
        "$dtId": "mockElementTwinId1",
        "$metadata": Object {
          "$model": "",
        },
        "InFlow": 60,
        "OutFlow": 35,
      },
    },
  },
  SceneVisual {
    "behaviors": Array [
      Object {
        "datasources": Array [
          Object {
            "elementIDs": Array [
              "MockElementId1",
              "MockElementId2",
              "MockElementId3",
            ],
            "type": "ElementTwinToObjectMappingDataSource",
          },
        ],
        "displayName": "Mock behavior 2",
        "id": "mockBehavior2",
        "visuals": Array [
          Object {
            "expressionType": "CategoricalValues",
            "objectIDs": Object {
              "expression": "N/A",
            },
            "type": "ExpressionRangeVisual",
            "valueExpression": "PrimaryTwin.Temperature",
            "valueRanges": Array [
              Object {
                "id": "valuerange1",
                "values": Array [
                  true,
                ],
                "visual": Object {
                  "color": "#d6940f",
                  "iconName": "Frigid",
                  "labelExpression": "Temp too hot",
                },
              },
            ],
          },
          Object {
            "expressionType": "NumericRange",
            "objectIDs": Object {
              "expression": "N/A",
            },
            "type": "ExpressionRangeVisual",
            "valueExpression": "PrimaryTwin.Temperature",
            "valueRanges": Array [
              Object {
                "id": "valuerange1",
                "values": Array [
                  0,
                  100,
                ],
                "visual": Object {
                  "color": "blue",
                },
              },
            ],
          },
        ],
      },
    ],
    "coloredMeshItems": undefined,
    "element": Object {
      "displayName": "MockElement3",
      "id": "MockElementId3",
      "objectIDs": Array [
        "elementObject2",
        "elementObject3",
      ],
      "primaryTwinID": "mockElementTwinId3",
      "twinAliases": Object {
        "TestAlias1": "aliasedTwinId1",
        "TestAlias2": "aliasedTwinId2",
      },
      "type": "TwinToObjectMapping",
    },
    "twins": Object {
      "PrimaryTwin": Object {
        "$dtId": "mockElementTwinId3",
        "$metadata": Object {
          "$model": "",
        },
        "Temperature": 99,
      },
      "TestAlias1": Object {
        "$dtId": "aliasedTwinId1",
        "$metadata": Object {
          "$model": "",
        },
        "InFlow": 60,
        "OutFlow": 35,
      },
      "TestAlias2": Object {
        "$dtId": "aliasedTwinId1",
        "$metadata": Object {
          "$model": "",
        },
        "InFlow": 60,
        "OutFlow": 35,
      },
    },
  },
]
`;
